From e78bc27055f77a98929bef67b4751b6a5b781ef8 Mon Sep 17 00:00:00 2001
Message-Id: <e78bc27055f77a98929bef67b4751b6a5b781ef8.1651164999.git.keller@fs-net.de>
From: Suresh Kumar Gautam <sgkuma@codeaurora.org>
Date: Mon, 6 Nov 2017 15:25:36 +0530
Subject: [PATCH 1/2] Increased delay for disconnecting client to 1 sec,as 10ms
 is too aggressive for stations. Cancel disconnect timer from Deauth/Disassoc
 Event handler.

---
 src/ap/ieee802_1x.c |  2 +-
 src/ap/sta_info.c   | 22 ++++++++++++++++++++--
 2 files changed, 21 insertions(+), 3 deletions(-)

diff --git a/src/ap/ieee802_1x.c b/src/ap/ieee802_1x.c
index 34637f3..33b6b3f 100644
--- a/src/ap/ieee802_1x.c
+++ b/src/ap/ieee802_1x.c
@@ -2599,7 +2599,7 @@ int ieee802_1x_eapol_tx_status(struct hostapd_data *hapd, struct sta_info *sta,
 	    ap_sta_pending_delayed_1x_auth_fail_disconnect(hapd, sta)) {
 		wpa_printf(MSG_DEBUG,
 			   "WPS: Indicate EAP completion on ACK for EAP-Failure");
-		hostapd_wps_eap_completed(hapd);
+		//hostapd_wps_eap_completed(hapd);
 	}
 #endif /* CONFIG_WPS */
 
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 51d7884..7141ecf 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -1369,6 +1369,15 @@ void ap_sta_deauth_cb(struct hostapd_data *hapd, struct sta_info *sta)
 	sta->flags &= ~WLAN_STA_PENDING_DEAUTH_CB;
 	eloop_cancel_timeout(ap_sta_deauth_cb_timeout, hapd, sta);
 	ap_sta_deauth_cb_timeout(hapd, sta);
+	if (eloop_cancel_timeout(ap_sta_delayed_1x_auth_fail_cb, hapd, sta) > 0)
+	{
+		wpa_printf(MSG_DEBUG,
+			   "%s: Removed ap_sta_delayed_1x_auth_fail_cb timeout for "
+			   MACSTR,
+			   hapd->conf->iface, MAC2STR(sta->addr));
+		if (sta->flags & WLAN_STA_WPS)
+			hostapd_wps_eap_completed(hapd);
+	}
 }
 
 
@@ -1381,6 +1390,15 @@ void ap_sta_disassoc_cb(struct hostapd_data *hapd, struct sta_info *sta)
 	sta->flags &= ~WLAN_STA_PENDING_DISASSOC_CB;
 	eloop_cancel_timeout(ap_sta_disassoc_cb_timeout, hapd, sta);
 	ap_sta_disassoc_cb_timeout(hapd, sta);
+	if (eloop_cancel_timeout(ap_sta_delayed_1x_auth_fail_cb, hapd, sta) > 0)
+	{
+		wpa_printf(MSG_DEBUG,
+			   "%s: Removed ap_sta_delayed_1x_auth_fail_cb timeout for "
+			   MACSTR,
+			   hapd->conf->iface, MAC2STR(sta->addr));
+		if (sta->flags & WLAN_STA_WPS)
+			hostapd_wps_eap_completed(hapd);
+	}
 }
 
 
@@ -1467,7 +1485,7 @@ void ap_sta_delayed_1x_auth_fail_disconnect(struct hostapd_data *hapd,
 {
 	wpa_dbg(hapd->msg_ctx, MSG_DEBUG,
 		"IEEE 802.1X: Force disconnection of " MACSTR
-		" after EAP-Failure in 10 ms", MAC2STR(sta->addr));
+		" after EAP-Failure in 1 sec", MAC2STR(sta->addr));
 
 	/*
 	 * Add a small sleep to increase likelihood of previously requested
@@ -1475,7 +1493,7 @@ void ap_sta_delayed_1x_auth_fail_disconnect(struct hostapd_data *hapd,
 	 * operations.
 	 */
 	eloop_cancel_timeout(ap_sta_delayed_1x_auth_fail_cb, hapd, sta);
-	eloop_register_timeout(0, 10000, ap_sta_delayed_1x_auth_fail_cb,
+	eloop_register_timeout(1, 10000, ap_sta_delayed_1x_auth_fail_cb,
 			       hapd, sta);
 }
 
-- 
2.14.5

